pkg/engine/resourcecache/k8sresource/k8sresource.go

```diff
@@ -0,0 +1,184 @@
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ResourceLoader) AddEntry(entry *v2alpha1.CachedContextEntry) error {
+	if entry.Spec.Resource == nil {
+		err := fmt.Errorf("Invalid object provided")
+		r.logger.Error(err, "")
+		return err
+	}
+	rc := entry.Spec.Resource
+	resource := schema.GroupVersionResource{
+		Group:    rc.Group,
+		Version:  rc.Version,
+		Resource: rc.Kind,
+	}
+	key := getKeyForResourceEntry(resource, rc.Namespace)
+	ent, err := r.createGenericListerForResource(resource, rc.Namespace)
+	if err != nil {
+		return err
+	}
+	r.logger.V(2).Info("key", key, "entry", ent)
+	ok := r.cache.Add(key, ent)
+	if !ok {
+		err := fmt.Errorf("failed to create cache entry key=%s", key)
+		r.logger.Error(err, "")
+		return err
+	}
+	r.logger.V(2).Info("successfully created cache entry")
+	return nil
+}
+
+func (r *ResourceLoader) Get(rc *kyvernov1.ResourceCache) (interface{}, error) {
+	if rc.Resource == nil {
+		return nil, fmt.Errorf("resource not found")
+	}
+	resource := schema.GroupVersionResource{
+		Group:    rc.Resource.Group,
+		Version:  rc.Resource.Version,
+		Resource: rc.Resource.Kind,
+	}
+	key := getKeyForResourceEntry(resource, rc.Resource.Namespace)
+	entry, ok := r.cache.Get(key)
+	if !ok {
+		err := fmt.Errorf("failed to fetch entry key=%s", key)
+		r.logger.Error(err, "")
+		return nil, err
+	}
+	r.logger.V(2).Info("key", key, "entry", entry)
+	r.logger.V(2).Info("successfully fetched cache entry")
+	return entry.Get()
+}
+
+func (r *ResourceLoader) Delete(entry *v2alpha1.CachedContextEntry) error {
+	if entry.Spec.Resource == nil {
+		return fmt.Errorf("invalid object provided")
+	}
+	rc := entry.Spec.Resource
+	resource := schema.GroupVersionResource{
+		Group:    rc.Group,
+		Version:  rc.Version,
+		Resource: rc.Kind,
+	}
+	key := getKeyForResourceEntry(resource, rc.Namespace)
+	ok := r.cache.Delete(key)
+	if !ok {
+		err := fmt.Errorf("failed to delete k8s object entry")
+		r.logger.Error(err, "")
+		return err
+	}
+	r.logger.V(2).Info("successfully deleted cache entry")
+	return nil
+}
+
+func (r *ResourceLoader) createGenericListerForResource(resource schema.GroupVersionResource, namespace string) (cache.ResourceEntry, error) {
+	informer := dynamicinformer.NewFilteredDynamicInformer(r.client, resource, namespace, resyncPeriod, k8scache.Indexers{k8scache.NamespaceIndex: k8scache.MetaNamespaceIndexFunc}, nil)
+	watchErrHandler := NewWatchErrorHandler(r.logger, resource, namespace)
+	err := informer.Informer().SetWatchErrorHandler(watchErrHandler.WatchErrorHandlerFunction())
+	if err != nil {
+		r.logger.Error(err, "failed to add watch error handler")
+		return nil, err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	go informer.Informer().Run(ctx.Done())
+	if !k8scache.WaitForCacheSync(ctx.Done(), informer.Informer().HasSynced) {
+		cancel()
+		err := errors.New("resource informer cache failed to sync")
+		r.logger.Error(err, "")
+		return nil, err
+	}
```

Comment:
@KhaledEmaraDev I dont think there is an event handler for this besides watch error handler. I am already waiting for cache to sync and it will sync irrespective of the fact that there is an error while listing/watching.

Doc says that WatchErrorHandler is called when there is an error while listing, and this is where the custom logic is supossed to be added
````golang
type SharedInformer interface {
    ...

    // The WatchErrorHandler is called whenever ListAndWatch drops the
    // connection with an error. After calling this handler, the informer
    // will backoff and retry.
    //
    // The default implementation looks at the error type and tries to log
    // the error message at an appropriate level.
    //
    // There's only one handler, so if you call this multiple times, last one
    // wins; calling after the informer has been started returns an error.
    //
    // The handler is intended for visibility, not to e.g. pause the consumers.
    // The handler should return quickly - any expensive processing should be
    // offloaded.
    SetWatchErrorHandler(handler WatchErrorHandler) error
}
```